src/wasm-type.h

```diff
@@ -164,12 +190,161 @@ struct Signature {
   }
   bool operator!=(const Signature& other) const { return !(*this == other); }
   bool operator<(const Signature& other) const;
+  std::string toString() const;
+};
+
+struct Field {
+  Type type;
+  bool mutable_;
+  Field(Type type, bool mutable_ = false) : type(type), mutable_(mutable_) {}
+  bool operator==(const Field& other) const {
+    return type == other.type && mutable_ == other.mutable_;
+  }
+  bool operator!=(const Field& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+typedef std::vector<Field> FieldList;
+
+struct Struct {
+  FieldList fields;
+  Struct(const Struct& other) : fields(other.fields) {}
+  Struct(FieldList fields) : fields(fields) {}
+  bool operator==(const Struct& other) const { return fields == other.fields; }
+  bool operator!=(const Struct& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+struct Array {
+  Field element;
+  Array(const Array& other) : element(other.element) {}
+  Array(Field element) : element(element) {}
+  bool operator==(const Array& other) const { return element == other.element; }
+  bool operator!=(const Array& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+union TypeDef {
+  enum Kind { TupleKind, SignatureKind, StructKind, ArrayKind };
+
+  struct TupleDef {
+    Kind kind;
+    Tuple tuple;
+  } tupleDef;
+  struct SignatureDef {
+    Kind kind;
+    Signature signature;
+    bool nullable;
+  } signatureDef;
+  struct StructDef {
+    Kind kind;
+    Struct struct_;
+    bool nullable;
+  } structDef;
+  struct ArrayDef {
+    Kind kind;
+    Array array;
+    bool nullable;
+  } arrayDef;
+
+  TypeDef(Tuple tuple) : tupleDef{TupleKind, tuple} {}
+  TypeDef(Signature signature, bool nullable)
+    : signatureDef{SignatureKind, signature, nullable} {}
+  TypeDef(Struct struct_, bool nullable)
+    : structDef{StructKind, struct_, nullable} {}
+  TypeDef(Array array, bool nullable) : arrayDef{ArrayKind, array, nullable} {}
+  TypeDef(const TypeDef& other) {
```

Comment:
You'll want to implement the copy assignment operator for this too

Reply:
Currently trying to do this, and came up with

```cpp
  TypeDef& operator=(const TypeDef& other) {
    if (&other != this) {
      kind = other.kind;
      switch (kind) {
        case TupleKind:
          new (&tupleDef) auto(other.tupleDef);
          break;
        case SignatureKind:
          new (&signatureDef) auto(other.signatureDef);
          break;
        case StructKind:
          new (&structDef) auto(other.structDef);
          break;
        case ArrayKind:
          new (&arrayDef) auto(other.arrayDef);
          break;
      }
    }
    return *this;
  }
```

I guess I should also implement copy assignment for `Tuple`, `Signature`, `Struct` and `Array` instead to make this worthwhile, that correct?

Reply:
The assignment operator is not quite right, you need to destroy the previous value before doing placement new.

> I guess I should also implement copy assignment for Tuple, Signature, Struct and Array instead to make this worthwhile, that correct?

No, generally you can use the default copy constructor/assignment operator, just in this case you can't because of the union.

Reply:
Oh, of course, thanks! Here's [the diff in isolation](https://github.com/WebAssembly/binaryen/pull/3012/commits/e2f328174b48cba86b864d5d5778f5c2c85d8b7c) :)

Reply:
Still not quite right, I think. :-) You need to destroy the type that was there before, then update the kind and construct the new type. It's all very tedious, but that's C++ for you!

Reply:
Hmm, so, just to make sure I understand correctly, what you are saying is that I instead need to destroy the entire thing with essentially `this.~TypeDef()`, and then set `kind` and do new placement of the concrete `xyDef`? I'd assume that this is safer in case `TypeDef` is ever extended with additional fields implying a destructor, while currently it doesn't matter functionally because the only other field is `kind`, which doesn't need to be destructed?

Also, I appreciate all the C++ you guys are teaching me, enjoying it :)

Reply:
I was thinking something like this: https://godbolt.org/z/G34z3f. There may be a simpler way to do this, but this is how I've seen it before.

Reply:
Whoops, yeah I'm setting `kind` too early. Completely missed that in between all the syntax :)

For completeness, what I was thinking above is something like

```cpp
  TypeDef& operator=(const TypeDef& other) {
    if (&other != this) {
      (*this).~TypeDef();
      new (this) auto(other);
    }
    return *this;
  }
```

so the existing constructor and destructor (already doing the switches and error handling) are reused without repeating their code in the copy assignment implementation. Is this pattern a bad idea?

Reply:
Ah, I see what you mean. Good idea!